[SPARK-17032][SQL] Add test cases for methods in ParserUtils.

jiangxb1987 opened the pull request:
## What changes were proposed in this pull request?

Currently methods in `ParserUtils` are tested indirectly, we should add test cases in `ParserUtilsSuite` to verify their integrity directly.
## How was this patch tested?

New test cases in `ParserUtilsSuite`

jiangxb1987 committed:
add testcases for ParserUtils.

SparkQA commented:
**[Test build #63680 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63680/consoleFull)** for PR 14620 at commit [`13891b6`](https://github.com/apache/spark/commit/13891b6a23b204b531e6a7011e1906bd83c685dd).

HyukjinKwon commented:
(Maybe you meant `SPARK-17032` in the title)

SparkQA commented:
**[Test build #63680 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63680/consoleFull)** for PR 14620 at commit [`13891b6`](https://github.com/apache/spark/commit/13891b6a23b204b531e6a7011e1906bd83c685dd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63680/
Test PASSed.

jiangxb1987 commented:
@HyukjinKwon Thank you for checking!

hvanhovell commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala:
```diff
@@ -61,5 +88,39 @@ class ParserUtilsSuite extends SparkFunSuite {
+  }
+
+  test("string") {
+    assert(string(showDbsContext.pattern) == "identifier_with_wildcards")
```
Could you also add a test for the other `string(node: TerminalNode)` code path?

hvanhovell commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala:
```diff
@@ -61,5 +88,39 @@ class ParserUtilsSuite extends SparkFunSuite {
 
-  // TODO: Add test cases for other methods in ParserUtils
+  test("command") {
+    assert(command(setConfContext) == "set example.setting.name=setting.value")
```
We can merge the both the command(..) functions in ParserUtils. Could you do this as part of this PR?

jiangxb1987 replied:
I searched and found the function `command(stream: CharStream)` was never used, so I simply removed it.

hvanhovell commented:
@jiangxb1987 thanks for doing this. I left a few comments. Any chance that you also add tests for the remaining methods?

jiangxb1987 commented:
@hvanhovell Thank you for your comments! Will do it as possible as I can.

jiangxb1987 committed:
1. Add test cases for function string(node: TerminalNode);
2. Combine the command(..) function in ParserUtils.

SparkQA commented:
**[Test build #63725 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63725/consoleFull)** for PR 14620 at commit [`bf57b91`](https://github.com/apache/spark/commit/bf57b9148997a33dda2662b2f8a1b2ccbc2c111a).

SparkQA commented:
**[Test build #63725 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63725/consoleFull)** for PR 14620 at commit [`bf57b91`](https://github.com/apache/spark/commit/bf57b9148997a33dda2662b2f8a1b2ccbc2c111a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63725/
Test PASSed.

hvanhovell commented:
@jiangxb1987 are you also adding tests for the remaining `ParserUtils` methods?

jiangxb1987 commented:
@hvanhovell I'm working on it, will try to add test cases for the remaining methods:
1. `operationNotAllowed(message: String, ctx: ParserRuleContext)`;
2. `checkDuplicateKeys[T](keyPairs: Seq[(String, T)], ctx: ParserRuleContext)`;
3. `assert(f: => Boolean, message: String, ctx: ParserRuleContext)`;
4. `withOrigin[T](ctx: ParserRuleContext)(f: => T)`.

jiangxb1987 committed:
add more test cases for methods in ParserUtils.

SparkQA commented:
**[Test build #63834 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63834/consoleFull)** for PR 14620 at commit [`a3068a7`](https://github.com/apache/spark/commit/a3068a726f53984242bc024fbe1fca597f2885c6).

jiangxb1987 commented:
I've added test cases for all the methods in `ParserUtils`, and made two small changes in the `ParserUtils` code:
1. Deleted function `command(stream: CharStream)`, merge both `command` method into `command(ctx: ParserRuleContext)`;
2. In method `position(token: Token)` the input param `token` could be `null`, so I added branch statement to handle the `token == null` condition.

SparkQA commented:
**[Test build #63834 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63834/consoleFull)** for PR 14620 at commit [`a3068a7`](https://github.com/apache/spark/commit/a3068a726f53984242bc024fbe1fca597f2885c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63834/
Test PASSed.

hvanhovell commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala:
```diff
@@ -61,5 +102,84 @@ class ParserUtilsSuite extends SparkFunSuite {
+
+  test("withOrigin") {
+    val ctx = createDbContext.locationSpec
+    val location = withOrigin(ctx) {
```
You also need to test the actual origin, use `org.apache.spark.sql.catalyst.trees.CurrentOrigin.get()` for this.

jiangxb1987 committed:
test the actual origin in withOrigin method.

SparkQA commented:
**[Test build #63856 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63856/consoleFull)** for PR 14620 at commit [`0a8f21e`](https://github.com/apache/spark/commit/0a8f21e4283e6db7cbd745bd4a30af0eea1f9282).

SparkQA commented:
**[Test build #63856 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63856/consoleFull)** for PR 14620 at commit [`0a8f21e`](https://github.com/apache/spark/commit/0a8f21e4283e6db7cbd745bd4a30af0eea1f9282).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63856/
Test PASSed.

hvanhovell commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParserUtils.scala:
```diff
@@ -74,7 +70,11 @@ object ParserUtils {
   /** Get the origin (line and position) of the token. */
   def position(token: Token): Origin = {
-    Origin(Option(token.getLine), Option(token.getCharPositionInLine))
+    if (token != null) {
```
how about:
```scala
val opt = Option(token)
Origin(opt.map(_.getLine), opt.map(_.getCharPositionInLine))
```

hvanhovell commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala:
```diff
@@ -61,5 +102,87 @@ class ParserUtilsSuite extends SparkFunSuite {
+  test("withOrigin") {
+    val ctx = createDbContext.locationSpec
+    val current = CurrentOrigin.get
+    val location = withOrigin(ctx) {
```
you also need to check if the origin of location to make this a complete tests, e.g.:
```scala
val (location, origin) = withOrigin(ctx) {
  (string(ctx.string), CurrentOrigin.get)
}
assert(origin, Origin(Some(2), Some(37)))
```

jiangxb1987 committed:
refactor code.

SparkQA commented:
**[Test build #63888 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63888/consoleFull)** for PR 14620 at commit [`36e049a`](https://github.com/apache/spark/commit/36e049a8fdafb3d311029b292e7d2de5efce4c6a).

SparkQA commented:
**[Test build #63888 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63888/consoleFull)** for PR 14620 at commit [`36e049a`](https://github.com/apache/spark/commit/36e049a8fdafb3d311029b292e7d2de5efce4c6a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63888/
Test PASSed.

hvanhovell commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/ParserUtilsSuite.scala:
```diff
@@ -61,5 +102,88 @@ class ParserUtilsSuite extends SparkFunSuite {
+    assert(position(emptyContext.stop) == Origin(None, None))
+  }
+
+  test("assert") {
```
We renamed assert into validate. So you will have to update. Sorry about that.

jiangxb1987 committed:
rename assert method to validate.

SparkQA commented:
**[Test build #63912 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63912/consoleFull)** for PR 14620 at commit [`11bcb1d`](https://github.com/apache/spark/commit/11bcb1d3f25bfc95878f8e4f3cf2270399987959).

hvanhovell commented:
LGTM - pending Jenkins.

SparkQA commented:
**[Test build #63912 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/63912/consoleFull)** for PR 14620 at commit [`11bcb1d`](https://github.com/apache/spark/commit/11bcb1d3f25bfc95878f8e4f3cf2270399987959).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/63912/
Test PASSed.

hvanhovell commented:
Merging to master. Thanks!

jiangxb1987 commented:
@hvanhovell Thank you for your help!
